[PNP-5604] Migrate html publication

andysellick opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What
Migrate HTML publication pages from `government-frontend`.

Example HTML publication pages:

- example [live](https://www.gov.uk/government/publications/car-show-me-tell-me-vehicle-safety-questions/car-show-me-tell-me-vehicle-safety-questions) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/car-show-me-tell-me-vehicle-safety-questions/car-show-me-tell-me-vehicle-safety-questions)
- very simple example [live](https://www.gov.uk/government/publications/application-to-naturalise-as-a-british-citizen-form-an/apply-online-form-an) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/application-to-naturalise-as-a-british-citizen-form-an/apply-online-form-an)
- very long contents list [live](https://www.gov.uk/government/publications/pace-code-c-2019/pace-code-c-2019-accessible) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/pace-code-c-2019/pace-code-c-2019-accessible)
- published by multiple organisations [live](https://www.gov.uk/government/publications/lowland-grazing-offer-countryside-stewardship/lowland-grazing-offer-what-you-need-to-know) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/lowland-grazing-offer-countryside-stewardship/lowland-grazing-offer-what-you-need-to-know)
- has 'applies to England' [live](https://www.gov.uk/government/publications/national-curriculum-in-england-english-programmes-of-study/national-curriculum-in-england-english-programmes-of-study) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/national-curriculum-in-england-english-programmes-of-study/national-curriculum-in-england-english-programmes-of-study)
- includes 'this guidance is issued by' [live](https://www.gov.uk/government/publications/fire-safety-england-regulations-2022-fire-door-guidance/fire-safety-england-regulations-2022-fire-door-guidance) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/fire-safety-england-regulations-2022-fire-door-guidance/fire-safety-england-regulations-2022-fire-door-guidance)
- published under a previous government [live](https://www.gov.uk/government/publications/apprentice-welcome-letter/this-content-has-been-removed-as-it-is-out-of-date) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/apprentice-welcome-letter/this-content-has-been-removed-as-it-is-out-of-date)
- withdrawn [live](https://www.gov.uk/government/publications/apprentice-welcome-letter/this-content-has-been-removed-as-it-is-out-of-date) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/apprentice-welcome-letter/this-content-has-been-removed-as-it-is-out-of-date)
- RTL language [live](https://www.gov.uk/government/publications/israel-and-the-occupied-palestinian-territories-country-of-concern/69c84613-51f7-49b8-9953-b37263c9f14b) / [review app](https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/israel-and-the-occupied-palestinian-territories-country-of-concern/69c84613-51f7-49b8-9953-b37263c9f14b)
- another RTL [live](https://www.gov.uk/government/publications/saudi-arabia-country-of-concern--2/968738)

Of type:

- [policy paper](https://www.gov.uk/government/publications/dvla-privacy-policy/dvla-privacy-policy)
- [guidance](https://www.gov.uk/government/publications/digital-id-scheme-explainer/digital-id-scheme-explainer)

Most popular HTML publications:

- [example 1](https://www.gov.uk/government/publications/car-show-me-tell-me-vehicle-safety-questions/car-show-me-tell-me-vehicle-safety-questions)
- [example 2](https://www.gov.uk/government/publications/tax-and-tax-credit-rates-and-thresholds-for-2017-18/tax-and-tax-credit-rates-and-thresholds-for-2017-18)
- [example 3](https://www.gov.uk/government/publications/rates-and-allowances-national-insurance-contributions/rates-and-allowances-national-insurance-contributions)
- [example 4](https://www.gov.uk/government/publications/rates-and-allowances-income-tax/income-tax-rates-and-allowances-current-and-past)
- [example 5](https://www.gov.uk/government/publications/income-tax-personal-allowance-and-basic-rate-limit-for-2017-to-2018/)
- [example 6](https://www.gov.uk/government/publications/recognising-the-terrorist-threat/recognising-the-terrorist-threat)
- [example 7](https://www.gov.uk/government/publications/tax-and-tax-credit-rates-and-thresholds-for-2016-17/tax-and-tax-credit-rates-and-thresholds-for-2016-17)
- [example 8](https://www.gov.uk/government/publications/advisory-fuel-rates/advisory-fuel-rates-from-1-march-2016)
- [example 9](https://www.gov.uk/government/publications/apprenticeship-levy-how-it-will-work/apprenticeship-levy-how-it-will-work)
- [example 10](https://www.gov.uk/government/publications/financial-sanctions-consolidated-list-of-targets/consolidated-list-of-targets)

More RTL HTML publications:



### Todo:

- [x] include the change in https://github.com/alphagov/government-frontend/pull/3860
- [x] decide how much space there should be at the top with the govuk-main-wrapper class

## Why
Part of the app consolidation work.

## Visual changes

Trello card: https://trello.com/c/w93U3be9/782-move-html-publication-from-government-frontend-to-frontend, [Jira issue PNP-5604](https://gov-uk.atlassian.net/browse/PNP-5604)

andysellick committed:
Add HTML publication document examples

andysellick committed:
Add placeholder routes and methods

andysellick committed:
Add request tests

- At this point we have a basic controller / view / route, so the basic request tests we add for all migrated
  routes will pass.

andysellick committed:
Add html publication CSS

Audit trail:
https://github.com/alphagov/government-frontend/blob/a38c7617449e4734fbc93a2bba35cc4c2c3bca79/app/assets/stylesheets/views/_html-publication.scss

andysellick committed:
Copy html publication view from government-frontend

- straight copy, no changes
- following commits will update the view to make it work properly

Audit trail:
https://github.com/alphagov/government-frontend/blob/a38c7617449e4734fbc93a2bba35cc4c2c3bca79/app/views/content_items/html_publication.html.erb

andysellick committed:
Fix lint errors in html publication document show template

- ran `bundle exec erb_lint --lint-all --autocorrect` to fix the errors

andysellick committed:
Update headers in html publication template

- include CSV preview page view styles for the organisation logos, both pages use the same styling

andysellick committed:
Replace @content_item with content_item

andysellick committed:
Create HtmlPublication model

- and empty model spec

andysellick committed:
Add HtmlPublicationPresenter

- and test, both blank for now

andysellick committed:
Basic HTML publication controller

andysellick committed:
Add hide_from_search_engines method

- we have to add a robots noindex tag to pages that are children of a list of specific pages in the publication presenter
- extends model to include parent

andysellick committed:
Add HTML publication header

- adds blue box containing document type, title and updated/published date

andysellick committed:
Add contents list

- has involved a change to the content item to add headers which at time of writing isn't live or included in any examples

andysellick committed:
Add govspeak body content

- rewrite CSS to reduce complexity (no need for .html-publication namespacing as Sass files are imported to individual views)
- include sticky positioning for contents list and right to left layout
- remove simple_header as not a thing

andysellick committed:
Add integration test

- omits create random doc and visit test
- omits direction-rtl test as this is covered in specific locale system tests
- audit trail: https://github.com/alphagov/government-frontend/blob/2a1927a3a8100374593aa708525d928a7e3c51a6/test/integration/html_publication_test.rb

andysellick committed:
Restore breadcrumbs

- breadcrumbs relies upon services.rb, which is a bit odd and will be refactored in a following PR
- fix a related spacing issue where the gap between breadcrumbs and org logos was too big
- this has also revealed that on pages with the prime ministers office org logo the current site doesn't show the left hand border on the org logo, because the component is being passed the wrong brand value (executive office instead of prime ministers office) so while this is a visual change it appears to be correcting a bug

KludgeKML commented on config/routes.rb:
```diff
@@ -209,6 +209,10 @@
     get "*path", to: "gone#show"
   end
 
+  constraints FullPathFormatRoutingConstraint.new("html_publication") do
```
If it doesn't cause the thing to break, can you move this block above landing page? (I'm trying to keep things in alphabetical order as much as possible - obviously we have to have all the fixed routes, then all the FormatRoutingConstraints, then all the FullPathFormatRoutingConstraints, but within each group I'm trying to sort alphabetically as a second sort.

KludgeKML commented on config/locales/ar.yml:
```diff
@@ -78,6 +78,502 @@ ar:
     bank_holiday_benefits_html:
```
I think we might have copied too many keys in this commit? They don't look strictly limited to html_publications keys.

andysellick replied:
Which keys are you referring to? I don't think line 78 was added by this PR?

KludgeKML commented on app/views/html_publication/show.html.erb:
```diff
@@ -34,11 +34,11 @@
   <%= render "govuk_publishing_components/components/heading", {
     text: content_item.title,
-    context: I18n.t("content_item.schema_name.#{content_item.format_sub_type}", count: 1),
+    context: I18n.t("common.content_item.schema_name.#{@presenter.format_sub_type}", count: 1),
```
I think this might need to be changed (in general I think we haven't explained what we're doing with locale keys, so this isn't on you).

KludgeKML commented on spec/models/html_publication_spec.rb:
```diff
@@ -66,4 +66,35 @@
       expect(html_publication.withdrawn_explanation).to eq("because of a reason")
     end
   end
+
```
Again, there's a shared example for this: https://github.com/alphagov/frontend/blob/main/spec/models/call_for_evidence_spec.rb#L10

KludgeKML commented on app/models/html_publication.rb:
```diff
@@ -29,4 +29,8 @@ def copyright_year
     content_store_response.dig("details", "isbn")
   end
+
+  def govspeak_body
```
content items already have a `body` attribute this is just this.

andysellick replied:
Thanks, have removed.

KludgeKML commented on lib/services.rb:
```diff
@@ -0,0 +1,13 @@
+require "gds_api/content_store"
```
I think this isn't needed in this commit? (or possibly at all)

KludgeKML reviewed (changes requested):
A lot of comments, but most of them are simple fixes (or allow you to just remove a load of code). I think it would be nice if any commits where you bring in locale keys could be separate from any other changes and could you include the command you ran (Particularly the first one I feel is probably wrong, but hard to see what happened without knowing what you ran).

andysellick commented on app/models/html_publication.rb:
```diff
@@ -2,4 +2,8 @@ class HtmlPublication < ContentItem
     linked("parent").first
   end
+
+  def organisations
```
Ah great thanks, have removed.

andysellick commented on spec/models/html_publication_spec.rb:
```diff
@@ -6,4 +6,9 @@
     expect(html_publication.parent.content_id).to eq("8b19c238-54e3-4e27-b0d7-60f8e2a677c9")
   end
+
+  it "returns organisations" do
```
Removed 👍

andysellick commented on spec/models/html_publication_spec.rb:
```diff
@@ -45,4 +45,25 @@
     end
   end
+
+  it "is not withdrawn" do
```
Updated 👍

andysellick commented on app/views/html_publication/show.html.erb:
```diff
@@ -68,9 +68,9 @@
 
 <div id="contents">
   <div class="govuk-grid-row gem-print-columns-none">
-    <% if content_item.contents.any? %>
```
Thanks, have updated 👍

andysellick commented on app/views/html_publication/show.html.erb:
```diff
@@ -13,9 +13,7 @@
-<%
-  content_for :simple_header, true
-%>
+<% content_for :simple_header, true %>
```
Thanks, have removed 👍 

See also https://github.com/alphagov/frontend/pull/5086

andysellick committed:
Add translations

- command run was
- bundle exec rake "consolidation:copy_translation[content_item.metadata,common.metadata]"

andysellick committed:
Add guidance translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.guidance,formats.guidance.name]"

andysellick committed:
Add corporate report translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.corporate_report,formats.corporate_report.name]"

andysellick committed:
Add form translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.form,formats.form.name]"

andysellick committed:
Add foi_release translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.foi_release,formats.foi_release.name]"

andysellick committed:
Add promotional translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.promotional,formats.promotional.name]"

andysellick committed:
Add notice translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.notice,formats.notice.name]"

andysellick committed:
Add correspondence translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.correspondence,formats.correspondence.name]"

andysellick committed:
Add research translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.research,formats.research.name]"

andysellick committed:
Add official_statistics translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.official_statistics,formats.official_statistics.name]"

andysellick committed:
Add transparency translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.transparency,formats.transparency.name]"

andysellick committed:
Add standard translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.standard,formats.standard.name]"

andysellick committed:
Add statutory_guidance translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.statutory_guidance,formats.statutory_guidance.name]"

andysellick committed:
Add independent_report translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.independent_report,formats.independent_report.name]"

andysellick committed:
Add national_statistics translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.national_statistics,formats.national_statistics.name]"

andysellick committed:
Add policy_paper translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.policy_paper,formats.policy_paper.name]"

andysellick committed:
Add decision translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.decision,formats.decision.name]"

andysellick committed:
Add map translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.map,formats.map.name]"

andysellick committed:
Add regulation translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.regulation,formats.regulation.name]"

andysellick committed:
Add international_treaty translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.international_treaty,formats.international_treaty.name]"

andysellick committed:
Add impact_assessment translations

- bundle exec rake "consolidation:copy_translation[content_item.schema_name.impact_assessment,formats.impact_assessment.name]"

andysellick committed:
Fix right to left language display

- some of the content wasn't displaying right to left when the page was in a right to left language
- inverse header fixed by adding dir from the component wrapper
- same for notices
- back to top link requires a custom style
- (interestingly govspeak was already correct, it handles itself in this situation)
- ideally we'd get a better solution to this problem, but out of scope for this PR

andysellick commented:
@KludgeKML I've spotted a problem with this version of https://www.gov.uk/government/publications/dvla-privacy-policy/dvla-privacy-policy where the contents list should have numbers in it, but doesn't.

Live | This branch
---- | -----------
<img width="258" height="212" alt="Screenshot 2025-10-24 at 09 11 45" src="https://github.com/user-attachments/assets/ae60a2c1-4e1e-422b-9d73-324977700d35" /> | <img width="254" height="205" alt="Screenshot 2025-10-24 at 09 11 53" src="https://github.com/user-attachments/assets/700078de-c52b-43a3-99b6-d64af5214014" />

I think the problem is that the content item headers have had the number prefix stripped from them. In the content the first heading is `1. Introduction` but in the content item headers its just `Introduction` (the contents list component then detects these numbers and formats them for display automatically).

KludgeKML commented on app/views/html_publication/show.html.erb:
```diff
@@ -46,7 +46,14 @@
 
 <% if content_item.withdrawn? %>
-  <%= render "govuk_publishing_components/components/notice", content_item.withdrawal_notice_component %>
+  <% withdrawn_time_tag = tag.time(display_date(content_item.withdrawn_at), datetime: content_item.withdrawn_at) %>
```
Not relevant to this exact line, but could you do the localisation copying in a separate commit?

KludgeKML commented on app/models/html_publication.rb:
```diff
@@ -1,5 +1,6 @@
 class HtmlPublication < ContentItem
   include Political
+  include NationalApplicability
```
Shared example for this could be added at the same time:
https://github.com/alphagov/frontend/blob/main/spec/models/call_for_evidence_spec.rb#L10

andysellick committed:
Add print copyright translations

andysellick committed:
Add print copyright statement

- should appear after the contents list, but only on print version, normally hidden
- simplify the classes and markup used to achieve this from the original, only used here

andysellick commented on app/assets/images/open-government-licence-min.png:
Yes, done, I think.

andysellick committed:
Add shared history notice

- adds a notification if the page was published under a former government
- uses the existing political.rb that appears to have already been migrated across with another route

andysellick committed:
Add withdrawn section

- adds a notice component if the content has been withdrawn
- uses existing withdrawable concern

andysellick committed:
Add withdrawn section translations

- translations were problematic, don't seem to exist in government-frontend and using add_missing created a load of extra unneeded things, had to add them by hand in the end

andysellick committed:
Fix HTML Publication header blue box

- the inverse header used for this part of the page was recently updated to be light blue instead of dark blue by default
- but we're keeping the dark blue on these pages for now, so adding the option to the component to retain that

KludgeKML reviewed (approved):
LGTM!

andysellick committed:
Add organisation logos

- can have multiple organisations, see https://www.gov.uk/government/publications/lowland-grazing-offer-countryside-stewardship/lowland-grazing-offer-what-you-need-to-know
- but if more than one organisation, remove images from being displayed in organisation logo components
- government-frontend had a whole stack of logic for handling organisations and branding, see https://github.com/alphagov/government-frontend/blob/main/app/presenters/content_item/organisation_branding.rb
- but this wasn't needed in this context, a simpler solution was possible

andysellick committed:
Add notice of national applicability

- shows the devolved nations component if content has applicability to specific nations
- uses the existing national applicability concern

AshGDS reviewed (commented):
Looks good to me apart from the contents list differences which I believe you're taking care of already?

I also noticed that on https://govuk-frontend-app-pr-4950.herokuapp.com/government/publications/apprentice-welcome-letter/this-content-has-been-removed-as-it-is-out-of-date  the page title is different compared to the live site https://www.gov.uk/government/publications/apprentice-welcome-letter/this-content-has-been-removed-as-it-is-out-of-date

andysellick committed:
Fix HTML Publication page titles

- uses https://github.com/alphagov/frontend/blob/main/app/presenters/content_item_presenter.rb#L14-L16
